datadryad/dryad-repo
# Dash transfer -- with bitstreams

Update for the Dash transfer process to include bitstreams. Also includes the code to "finalize" a submission and send it to Merritt (somehow I thought that was merged earlier). Sorry this is so long!

See 

- https://github.com/CDL-Dryad/dryad-product-roadmap/issues/57
- https://github.com/CDL-Dryad/dryad-product-roadmap/issues/58

Random notes:

- This includes a small bugfix for S3 bucket names. It ensures that if you don't specify a bucket name, the automatically-generated name will not include a period. AWS discourages you from using periods in a bucket name, but doesn't forbid it. If the bucket name includes a period, the AWS presigned URLs will generate SSL certificate errors, so don't do this!
- Handling of HTTP Response streams. The API calls have been changed to first test for the presence of an error stream, and get the normal input stream when there is no error stream. This is incredibly helpful in actually seeing the contents of an error message when it exists.

To test:
1. Ensure your maven settings file has appropriate values for the variables introduced in #1480, as well as these new variables:
    - `default.dash.submissions.finalize` = If true, the submission will be submitted for processing by Merritt. If false, all transfer will be done *except* the final submission (useful for dev servers).
    -  `default.dash.submissions.delaySeconds` = number of seconds to delay after each data package is submitted, to not overwhelm the dash server (a good default is 5)
2. Find the handle of an item you want to transfer
3. To post a new item to Dash, call the curation tool from the command line, using the handle:
`/opt/dryad/bin/dspace curate -v -t transfertodash -i 10255/dryad.135814 -r -`
4. Login to the Dash server and view the item.
